grammar/abnf-grammar.txt

```diff
@@ -1045,16 +1052,17 @@ member-variable-declarations = *( identifier ":" type ( "," / ";" ) )
 
 member-function-declaration = function-declaration
 
-; A circuit declaration defines a circuit type,
-; as consisting of member variables and functions.
+; A circuit member constant declaration
+; as consisting of member constants, member variables, and member functions.
 ; To more simply accommodate the backward compatibility
 ; described for the rule `member-variable-declarations`,
 ; all the member variables must precede all the member functions;
 ; this may be relaxed after the backward compatibility is removed,
 ; allowing member variables and member functions to be intermixed.
 
 circuit-declaration = %s"circuit" identifier
-                      "{" [ member-variable-declarations ]
+                      "{" *member-constant-declaration
+                      [ member-variable-declarations ]
                       *member-function-declaration "}"
```

Comment:
I would change the grammar to be `*(const-decl / variable-decl / function)` instead. As it is, this complicates the parsing and makes the UX worse for users.

Reply:
At least for const-decl and variable-decl these would have to be a different rule. As in these cases specifying types are not optional.

Probably deserves it's own issue to discuss this topic as well as it would require a number of grammar changes not just in this section.

Reply:
> At least for const-decl and variable-decl these would have to be a different rule. As in these cases specifying types are not optional.

Not sure I follow how non-optional types affects whether you have to put all of the same kind contiguously or whether they can be interspersed randomly.

> Probably deserves it's own issue to discuss this topic as well as it would require a number of grammar changes not just in this section.

We could start with this just being error recovery for now by first parsing it and then checking whether they are contiguous and if they aren't, then emit an error for the incorrect order.

Reply:
Ahhh, I misunderstood that part then as well; my bad. I thought you were suggesting the re-names on top of the changes.

Reply:
Yes, that was the plan, i.e. having declarations of member constants, variables, and functions be freely interspersed. The current awkward grammar rule is a concession to backward compatibility: originally circuit member variables were comma-separated, but then we figured that, since we have member function declarations that are not comma-separated, it makes more sense to terminate member variables with semicolons. But we wanted to still allow commas for a while, resulting in an admittedly awkward rule that's error-prone to implement :)

But this also interacts with https://github.com/AleoHQ/leo/issues/1417, so that's part of why we've been holding off on that awkward grammar rule.

Reply:
Makes sense, I'll see if I can improve error recovery separately despite the grammar being awkward.

Reply:
Thanks. Good challenge for error recovery :)